upload-api/stores/agent/stream.js

```diff
@@ -0,0 +1,154 @@
+ *   disable: {}
+ * }>} Connection
+ *
+ * @typedef {object} Options
+ * @property {Connection} connection
+ * @property {string} name
+ * @property {string} [partitionKey]
+ * @property {{type:string}} [workflow]
+ * @property {{type:string}} [receipt]
+ *
+ * @typedef {object} Stream
+ * @property {Channel|null} channel
+ * @property {string} name
+ * @property {string} partitionKey
+ * @property {{type:string}} workflow
+ * @property {{type:string}} receipt
+ */
+
+/**
+ * @param {Options} options
+ * @returns {Stream}
+ */
+export const open = ({ connection, name, partitionKey, ...settings }) => ({
+  ...settings,
+  channel: connection.address
+    ? new Kinesis(connection.address)
+    : connection.channel ?? null,
+  name,
+  workflow: { ...defaults.workflow, ...settings.workflow },
+  receipt: { ...defaults.receipt, ...settings.receipt },
+  partitionKey: partitionKey ?? defaults.partitionKey,
+})
+
+/**
+ *
+ * @param {object} connection
+ * @param {Store.Store} connection.store
+ * @param {Stream} connection.stream
+ * @param {API.ParsedAgentMessage} message
+ * @returns {Promise<API.Result<API.Unit, Error>>}
+ */
+export const write = async (connection, message) => {
+  const { stream } = connection
+
+  try {
+    if (stream.channel) {
+      await stream.channel.putRecords({
+        Records: await assert(message, connection),
+        StreamName: connection.stream.name,
+      })
+    }
+    return { ok: {} }
+  } catch (cause) {
+    return { error: /** @type {Error} */ (cause) }
+  }
+}
+
+/**
+ * Iterates over all invocations and receipts and yields corresponding kinesis
+ * records.
+ *
+ * @param {API.ParsedAgentMessage} message
+ * @param {object} connection
+ * @param {Store.Store} connection.store
+ * @param {Stream} connection.stream
+ */
+export const assert = async (message, { stream, store }) => {
+  const records = []
+  for (const member of message.index) {
+    if (member.invocation) {
+      const { task, invocation, message } = member.invocation
+      records.push({
+        Data: UTF8.fromString(
+          JSON.stringify({
+            // This is bad naming but not worth a breaking change
+            carCid: message.toString(),
+            task: task.toString(),
+            value: {
+              att: invocation.capabilities,
+              aud: invocation.audience.did(),
+              iss: invocation.issuer.did(),
+              cid: invocation.cid.toString(),
+            },
+            ts: Date.now(),
+            type: stream.workflow.type,
+          })
+        ),
+        PartitionKey: stream.partitionKey,
+      })
+    }
+
+    if (member.receipt) {
+      const { task, receipt, message } = member.receipt
+      // Prior implementation used to resolve the invocation and include it's
+      // details the stream record. If invocation was not found it threw
+      // exception which would result in HTTP status 500.
+
+      // 🔬 Need to figure if there are consumers downstream that depend on
```

Comment:
IIRC it is currently needed. There are consumers that need the information of the invocation to track metrics. For instance, when we receive a `store/add` receipt with 'ok', we need the invocation info to know the `size` value to increment the size os bytes received by the service

I would wish we just attempt these changes in follow up, or we are going to make this quite complicated